src/queues/sqs.rs

```diff
@@ -0,0 +1,252 @@
+            .client
+            .receive_message(&self.receive_request)
+            .map(|result| result.messages.unwrap_or(Vec::new()))
+            .map(move |messages| {
+                messages
+                    .into_iter()
+                    .map(|message| {
+                        self.parse_message(message).unwrap_or_else(|error| {
+                            // At this point any parse errors are message-specific.
+                            // Log them but don't fail the broader call to receive,
+                            // because other messages might be fine.
+                            println!(
+                                "Queue error receiving from {}: {}",
+                                self.url,
+                                error.description()
+                            );
+                            Message::default()
+                        })
+                    })
+                    .collect()
+            })
+            .map_err(From::from);
+
+        Box::new(future)
+    }
+
+    fn delete(&'static self, message: Message) -> DeleteFuture {
+        let request = DeleteMessageRequest {
+            queue_url: self.url.to_string(),
+            receipt_handle: message.id,
+        };
+
+        let future = self.client.delete_message(&request).map_err(move |error| {
+            println!(
+                "Queue error deleting from {}: {}",
+                self.url,
+                error.description()
+            );
+            From::from(error)
+        });
+
+        Box::new(future)
+    }
+}
+
+impl Outgoing for Queue {
+    fn send(&'static self, notification: &Notification) -> SendFuture {
+        let mut request = SendMessageRequest::default();
+        request.message_body = match serde_json::to_string(notification) {
+            Ok(body) => body,
+            Err(error) => return Box::new(future::err(From::from(error))),
+        };
+        request.queue_url = self.url.to_string();
+
+        let future = self
+            .client
+            .send_message(&request)
+            .map(|result| result.message_id.map_or(String::from(""), |id| id.clone()))
+            .map_err(From::from);
+
+        Box::new(future)
+    }
+}
+
+impl From<ReceiveMessageError> for QueueError {
+    fn from(error: ReceiveMessageError) -> QueueError {
+        QueueError::new(format!("SQS ReceiveMessage error: {:?}", error))
+    }
+}
+
+impl From<SendMessageError> for QueueError {
+    fn from(error: SendMessageError) -> QueueError {
+        QueueError::new(format!("SQS SendMessage error: {:?}", error))
+    }
+}
+
+impl From<DeleteMessageError> for QueueError {
+    fn from(error: DeleteMessageError) -> QueueError {
+        QueueError::new(format!("SQS DeleteMessage error: {:?}", error))
+    }
+}
+
+impl From<JsonError> for QueueError {
+    fn from(error: JsonError) -> QueueError {
+        QueueError::new(format!("JSON error: {:?}", error))
+    }
+}
+
+enum SqsWrapper {
```

Comment:
I've gotta admit, I don't really understand what this is for.  Is there a way to explain it in a comment here, or does it depend too much on the details of the above compiler error you mentioned?

Reply:
Yep, will do.

Just to answer it here too, the obvious fix to the problem of `Debug` not being implemented (the compiler error mentioned above), would be just to do something like this:

```rust
impl Debug for SqsClient<CredentialsProvider, RequestDispatcher> {
    ...
}
```

But if you try to do that, you get this error:

```
error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
   --> src/queues/sqs.rs:217:1
    |
217 | impl Debug for SqsClient<CredentialsProvider, RequestDispatcher> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ impl doesn't use types inside crate
    |
    = note: the impl does not reference any types defined in this crate
    = note: define and implement a trait or new type instead
```

So in order to `Box` the SQS client, I needed a way to implement `Debug` but I wasn't allowed to do it directly. So I wrap it in a new type, proxying the methods that I want to call, and implementing `Debug` on the outer type.

But because there are actually two different concrete specialisations of the abstract `SqsClient` type (which is why I needed to `Box` it in the first place of course), I want to wrap both of those concrete types with the same implementation. Thus the wrapping type is an `enum` and I proxy to the appropriate inner type at runtime.

There may, probably are, better ways to do this I suspect. It was a home-baked solution and I didn't spend a great deal of time on it.

Reply:
...although interestingly, I finally got round to producing a cut-down example of the problem with a view to opening an issue, and it works okay there. So something else I was doing must have been causing it, will fix.